src/main.rs

```diff
@@ -532,42 +541,41 @@ async fn root_post(
         .await
         .map_err(|port_conflicts| redirect::port_conflicts(&port_conflicts).into_response())?;
 
-    // Create the new job and get an identifier.
-    let uuid = {
-        let mut lock = user.write().await;
+    // Create the new job.
+    let uuid = Uuid::new_v4();
+    {
+        let mut lock = JOBS.write().await;
 
-        if lock.data.job().is_some() {
+        if JOBS.read().await.by_user(&user).is_some() {
             return Err(Redirect::to("/").into_response());
         }
 
-        if jobs::Job::count() >= jobs {
+        if JOBS.read().await.i2j.len() >= jobs {
             return Err(redirect::too_many_workloads().into_response());
         }
 
-        let job = jobs::Job::new(command, workload_type, slug, wasm, toml, ports)?;
-        let uuid = job.uuid;
-        lock.data = Data::new(Some(job));
-        uuid
-    };
+        let job = Job::new(command, workload_type, slug, wasm, toml, ports)?;
+
+        let _ = lock.i2j.insert(uuid, RwLock::new(job).into());
+        let _ = lock.u2i.insert(user, uuid);
+    }
 
     // Set the job timeout.
-    let weak = Ref::downgrade(&user);
     _ = tokio::spawn(async move {
         sleep(ttl).await;
+        let mut lock = JOBS.write().await;
 
-        if let Some(user) = weak.upgrade() {
+        if lock.i2j.remove(&uuid).is_some() {
             debug!("timeout for: {}", uuid);
-            let mut lock = user.write().await;
-            if lock.data.job().as_ref().map(|j| j.uuid) == Some(uuid) {
-                lock.data.kill_job().await;
-            }
+        }
+
+        if lock.u2i.get(&user) == Some(&uuid) {
+            let _ = lock.u2i.remove(&user);
```

Comment:
using `let _ = lock.u2i.remove(&user);` and not killing the job will mean ports and docker containers will not be freed. the job will not be killed and we will run out of resources